### PR TITLE
Add ordered xds push

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

kind of cleanup. based on https://github.com/istio/istio/pull/48542

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
supersede #48519， if we think allocation is a big cost, we can make use of golang sync pool.

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -1007,33 +994,38 @@ func (conn *Connection) Watched(typeUrl string) *model.WatchedResource {
...
-		typeUrls.Insert(k)
+// orderedXdsPush pushes xds to the connection in order.
+func (s *DiscoveryServer) orderedXdsPush(
+	conn *Connection,
+	pushRequest *model.PushRequest,
+	push func(con *Connection, w *model.WatchedResource, req *model.PushRequest) error,
+) error {
+	// first push all known types, in order
+	for _, tp := range PushOrder {
+		conn.proxy.RLock()
```
Comment (ramaraochavali) [inline pilot/pkg/xds/ads.go:1005]: I think this has two issues
1. We are not ordering the entire PushOrder once as we used to do earlier. We are reading one typeURL at a time which may lead to inconsistencies. Not sure what impacts it will have
2. Locking/Releasing multiple times - not sure if it is slower than locking once and reading

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -1007,33 +994,38 @@ func (conn *Connection) Watched(typeUrl string) *model.WatchedResource {
...
 
-func orderWatchedResources(resources map[string]*model.WatchedResource) []*model.WatchedResource {
-	wr := make([]*model.WatchedResource, 0, len(resources))
-	// first add all known types, in order
-	for _, tp := range PushOrder {
-		if w, f := resources[tp]; f {
-			wr = append(wr, w)
+	// Then push any undeclared ordered types
+	conn.proxy.RLock()
+	resources := make([]*model.WatchedResource, 0, len(conn.proxy.WatchedResources))
```
Comment (ramaraochavali) [inline pilot/pkg/xds/ads.go:1017]: We are not avoiding this allocation with this approach

File: pilot/pkg/xds/ads.go [hunk 3]
```diff
@@ -782,15 +782,8 @@ func (s *DiscoveryServer) pushConnection(con *Connection, pushEv *Event) error {
 
 	// Send pushes to all generators
 	// Each Generator is responsible for determining if the push event requires a push
-	wrl, ignoreEvents := con.pushDetails()
-	for _, w := range wrl {
-		if err := s.pushXds(con, w, pushRequest); err != nil {
-			return err
-		}
-	}
-	if pushRequest.Full {
```
Comment (ramaraochavali) [inline pilot/pkg/xds/ads.go:791]: Why is this removed?

**Review by ramaraochavali — COMMENTED** [review]
I still think caching in Proxy performs better similar to what I have done in https://github.com/istio/istio/pull/48519 is better. May be we should create a benchmark test to see if what performs better - may be leave what we have as is if the benefit is negligible as the current code is simpler

**Review by hzxuzhonghu — COMMENTED** [review]
/close
